api/genericresource/resource_management.go

```diff
@@ -0,0 +1,194 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assigns GenericResources to a task by taking them from the
+// node's GenericResource list and storing them in the task's available list
+func Claim(nodeAvailable, taskAssigned *[]*api.GenericResource,
+	taskReservations []*api.GenericResource) error {
+	resSelected := make([]*api.GenericResource, 0)
+
+	for _, res := range taskReservations {
+		tr := res.GetDiscrete()
+		if tr == nil {
+			return newError("task should only hold Discrete type")
+		}
+
+		// Select the resources
+		nrs, err := selectNodeResources(nodeAvailable, tr)
+		if err != nil {
+			return err
+		}
+
+		resSelected = append(resSelected, nrs...)
+	}
+
+	ClaimResources(nodeAvailable, taskAssigned, resSelected)
+	return nil
+}
+
+// ClaimResources adds the specified resources to the task's list
+// and removes them from the node's generic resource list
+func ClaimResources(nodeAvailable, taskAssigned *[]*api.GenericResource,
+	resSelected []*api.GenericResource) {
+	*taskAssigned = append(*taskAssigned, resSelected...)
+	RemoveNodeResources(nodeAvailable, resSelected)
+}
+
+func selectNodeResources(nodeRes *[]*api.GenericResource,
+	tr *api.GenericDiscrete) (nrs []*api.GenericResource, err error) {
+	nrs = make([]*api.GenericResource, 0, tr.Value)
+
+loop:
+	for _, res := range *nodeRes {
+		if getKind(res) != tr.Kind {
+			continue
+		}
+
+		switch nr := res.Resource.(type) {
+		case *api.GenericResource_Discrete:
+			if nr.Discrete.Value >= tr.Value && tr.Value != 0 {
+				nrs = append(nrs, NewDiscrete(tr.Kind, tr.Value))
+			}
+
+			break loop
+		case *api.GenericResource_Str:
+			nrs = append(nrs, res)
+
+			if int64(len(nrs)) == tr.Value {
+				break loop
+			}
+		}
+	}
+
+	if len(nrs) == 0 {
+		return nil, newError("Not enough resource available for task reservations: %+v\n", tr)
+	}
+
+	return nrs, nil
+}
+
+// Reclaim adds the resources taken by the task to the node's store
+func Reclaim(nodeAvailable *[]*api.GenericResource, taskAssigned []*api.GenericResource) error {
+	// The node could have been updated
+	if nodeAvailable == nil {
```

Comment:
`Availability` was a term Swarmkit uses for node's administrative status, with value `Active, Pause, Drain`. Maybe choose another name for available resource.